# Prova 1 - 24.2

In [ ]:
def alg(n):
  if n == 1:
    return 1
  else:
    return alg(n - 1) + n ** 3

def alg_iter(n):
    S = 0
    for i in range(1, n + 1):  # Corrigido para iterar de 1 até n
        S += i ** 3  # Somamos i³, não n³
    return S

print(alg(4) == alg_iter(4))  # Deve retornar 1³ + 2³ + 3³ + 4³ = 100

True


In [ ]:
def visualize_matrix(M):
  for i in range(len(M)):
    for j in range(len(M[0])):
      print(M[i][j], end=' ')
    print()

def sum_matrices(A, B):
  C = [[0 for _ in range(len(A[0]))] for _ in range(len(A))]

  for i in range(len(A)):
    for j in range(len(A[0])):
      C[i][j] += A[i][j] + B[i][j]

  return C

A = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
B = [[10, 11, 12], [13, 14, 15], [16, 17, 19]]

C = sum_matrices(A, B)
visualize_matrix(C)

11 13 15 
17 19 21 
23 25 28 


In [ ]:
def find_missing_value(arr):
  n = len(arr) + 1
  tot_sum = n * (n + 1) // 2
  act_sum = sum(arr)
  return tot_sum - act_sum

arr = [1, 2, 4, 6, 3, 7, 8]
missing_value = find_missing_value(arr)
print("Missing value:", missing_value)

Missing value: 5


In [ ]:
def find_leaders(arr):
    leaders = []
    max_right = -float('inf')  # Inicializa com "infinito negativo"

    # Percorre da última posição até a primeira
    i = len(arr) - 1
    while i >= 0:
        if arr[i] >= max_right:
            leaders.append(arr[i])
            max_right = arr[i]
        i -= 1  # Move para o elemento anterior

    # Retorna a lista invertida para manter a ordem original
    return leaders[::-1]

# Exemplo de uso
arr = [16, 17, 4, 3, 5, 2]
print(find_leaders(arr))  # Saída: [17, 5, 2]

[17, 5, 2]


In [ ]:
def rotate(arr, r):
  aux = arr[-r::]
  print(aux)

def flatten(arr):
  aux = []

  for i in range(len(arr)):
    for j in range(len(arr[i])):
      aux.append(arr[i][j])

  return aux

print(flatten([[1, 2], [3, 4], [5]]))
#print(rotate([1, 2, 3, 4, 5], 2))

[1, 2, 3, 4, 5]


# Lista 1

## Chapter 2 - Levitin

### Questão 1)

- **Complexidade de Tempo**: Mede quantas operações um algoritmo executa em função da entrada $n$ (ex: $O(n)$, $O(n^2)$).  
- **Complexidade de Espaço**: Mede a quantidade de memória extra usada pelo algoritmo em função da entrada $n$.

### Questão 2)

- **Melhor Caso**: O desempenho do algoritmo quando executa a menor quantidade possível de operações (ex: entrada já ordenada em um algoritmo de ordenação).  
- **Caso Médio**: O desempenho esperado do algoritmo para uma entrada aleatória, considerando a distribuição de probabilidades dos casos possíveis.  
- **Pior Caso**: O desempenho máximo (mais lento) do algoritmo, ou seja, quando ele executa o maior número possível de operações.

### Questão 3)

Os problemas de usar **unidades de tempo (segundos)** para analisar o tempo de execução de algoritmos são:  

1. **Dependência do hardware**: O tempo de execução varia conforme o processador, memória e otimizações do sistema.  
2. **Variações externas**: Fatores como carga do sistema operacional e cache afetam o tempo medido.  
3. **Escalabilidade ruim**: Não permite comparar algoritmos de forma geral, pois diferentes máquinas podem ter tempos diferentes.  

**Estratégia mais adequada**  
A melhor abordagem é a **análise assintótica**, que usa a **notação Big-O** ($O(n)$, $O(\log n)$, etc.). Ela mede a quantidade de operações em função do tamanho da entrada, independentemente do hardware.

### Questão 4)

Se o tamanho do argumento **$n$** aumentar **4 vezes** (ou seja, substituir $n$ por $4n$), o crescimento das funções será:  

(a) **$\log_2 n$** → $\log_2 (4n) = \log_2 4 + \log_2 n = 2 + \log_2 n$ → **Aumenta em 2 unidades**.  

(b) **$\sqrt{n}$** → $\sqrt{4n} = 2\sqrt{n}$ → **Multiplica por 2**.  

(c) **$n$** → $4n$ → **Multiplica por 4**.  

(d) **$n^2$** → $(4n)^2 = 16n^2$ → **Multiplica por $4^2$**.  

(e) **$n^3$** → $(4n)^3 = 64n^3$ → **Multiplica por $4^3$**.  

(f) **$2^n$** → $2^{4n}$ → $\frac{2^{4n}}{2^n} = 2^{3n}$**Triplica o expoente**, ou seja, **cresce exponencialmente mais rápido**.

### Questão 5)

(a) O número de multiplicações do algoritmo de **Eliminação Gaussiana** é aproximadamente $\frac{1}{3} n^3$. Assim, a razão entre os tempos de execução para **$n = 1000$** e **$n = 500$** será:  

$ \frac{\frac{1}{3} (1000)^3}{\frac{1}{3} (500)^3} = \frac{1000^3}{500^3} = \left(\frac{1000}{500}\right)^3 = 2^3 = 8 $  

Ou seja, **o sistema com 1000 equações levará aproximadamente 8 vezes mais tempo** que o sistema com 500 equações.  

---

(b) Se o novo computador for **1000 vezes mais rápido**, isso significa que ele pode realizar **1000 vezes mais operações no mesmo tempo**. Queremos determinar o novo tamanho **$n_{\text{novo}}$** que pode ser resolvido no mesmo tempo que um sistema de tamanho **$n_{\text{antigo}}$** no computador antigo.  

Sabemos que o tempo de execução cresce como **$n^3$**, então precisamos resolver:  

$ (n_{\text{novo}})^3 = 1000 \cdot (n_{\text{antigo}})^3 $  

Tomando a raiz cúbica em ambos os lados:  

$ n_{\text{novo}} = \sqrt[3]{1000} \cdot n_{\text{antigo}} = 10 \cdot n_{\text{antigo}} $  

Isso significa que **o novo computador permitirá resolver sistemas 10 vezes maiores no mesmo tempo**.

### Questão 6)

As notações **Big-O ($O$), Ômega ($\Omega$) e Teta ($\Theta$)** são usadas para descrever a complexidade assintótica de algoritmos:  

- **$O(g(n))$ (Limite Superior - "no pior caso")**  
  Representa um limite superior para o crescimento da função. Formalmente, uma função $f(n)$ está em $O(g(n))$ se existirem constantes positivas $c$ e $n_0$ tais que:  
  $$ f(n) \leq c \cdot g(n), \quad \text{para todo } n \geq n_0 $$  
  **Exemplo:** Se um algoritmo tem complexidade $O(n^2)$, ele nunca crescerá mais rápido do que uma função proporcional a $n^2$ para entradas suficientemente grandes.  

- **$\Omega(g(n))$ (Limite Inferior - "no melhor caso")**  
  Representa um limite inferior para o crescimento da função. Formalmente, $f(n)$ está em $\Omega(g(n))$ se existirem constantes positivas $c$ e $n_0$ tais que:  
  $$ f(n) \geq c \cdot g(n), \quad \text{para todo } n \geq n_0 $$  
  **Exemplo:** Se um algoritmo tem complexidade $\Omega(n)$, ele sempre executará pelo menos um número de operações proporcional a $n$ para entradas suficientemente grandes.  

- **$\Theta(g(n))$ (Limite Exato - "crescimento exato")**  
  Indica que $f(n)$ cresce na mesma taxa que $g(n)$. Formalmente, $f(n)$ está em $\Theta(g(n))$ se existirem constantes positivas $c_1$, $c_2$ e $n_0$ tais que:  
  $$ c_1 \cdot g(n) \leq f(n) \leq c_2 \cdot g(n), \quad \text{para todo } n \geq n_0 $$  
  **Exemplo:** Se um algoritmo tem complexidade $\Theta(n^2)$, isso significa que ele cresce **exatamente** na ordem de $n^2$ (nem mais rápido, nem mais lento) para entradas grandes.

### Questão 7)

Para provar o teorema, precisamos mostrar que se $t_1(n) \in O(g_1(n))$ e $t_2(n) \in O(g_2(n))$, então:  

$$
t_1(n) + t_2(n) \in O(\max\{g_1(n), g_2(n)\})
$$

#### **Passo 1: Definições de Notação Big-O**  
Pela definição de Big-O, sabemos que:  

$$
t_1(n) \leq c_1 g_1(n), \quad \text{para todo } n \geq n_1
$$

$$
t_2(n) \leq c_2 g_2(n), \quad \text{para todo } n \geq n_2
$$

#### **Passo 2: Somando as desigualdades**  
Se tomarmos $n_0 = \max\{n_1, n_2\}$, então para todo $n \geq n_0$, temos:  

$$
t_1(n) + t_2(n) \leq c_1 g_1(n) + c_2 g_2(n)
$$

Como **$\max\{g_1(n), g_2(n)\}$** representa a função dominante entre $g_1(n)$ e $g_2(n)$, sabemos que ambas as funções são menores ou iguais a um múltiplo de $\max\{g_1(n), g_2(n)\}$, ou seja:  

$$
g_1(n) \leq \max\{g_1(n), g_2(n)\}, \quad g_2(n) \leq \max\{g_1(n), g_2(n)\}
$$

Portanto, podemos escrever:  

$$
t_1(n) + t_2(n) \leq c_1 \max\{g_1(n), g_2(n)\} + c_2 \max\{g_1(n), g_2(n)\}
$$

$$
t_1(n) + t_2(n) \leq (c_1 + c_2) \max\{g_1(n), g_2(n)\}
$$

#### **Passo 3: Conclusão**  
A última desigualdade mostra que $t_1(n) + t_2(n)$ está limitado superiormente por uma constante vezes $\max\{g_1(n), g_2(n)\}$ para todo $n \geq n_0$. Isso significa que:  

$$
t_1(n) + t_2(n) \in O(\max\{g_1(n), g_2(n)\})
$$  

### Questão 8)

Aqui estão as equações com as modificações solicitadas:

**(a) $n(n - 1), n^2$**

1. **Limite:**

$$
\lim_{n \to \infty} \frac{n(n - 1)}{n^2} = 1 - \frac{1}{n}
$$

O limite é 1, o que indica que $ n(n - 1) $ e $ n^2 $ têm o mesmo comportamento assintótico.

2. **Conclusão em termos de Big-O, Theta e Omega:**
   - Como o limite é uma constante diferente de zero, podemos concluir que:

$$
n(n - 1) \in \Theta(n^2)
$$

Isso significa que $ n(n - 1) $ e $ n^2 $ crescem na mesma taxa assintótica.

Além disso, já que o limite é 1 e não 0, isso implica também que:

$$
n(n - 1) \in O(n^2) \quad \text{e} \quad n^2 \in \Omega(n(n - 1))
$$

**(b) $ \log_2 n $ e $ \sqrt{n} $**

1. **Limite:**

$$
\lim_{n \to \infty} \frac{\log_2 n}{\sqrt{n}} = 0
$$

Como o limite é 0, isso indica que $ \log_2 n $ cresce mais lentamente que $ \sqrt{n} $.

2. **Conclusão em termos de Big-O, Theta e Omega:**

- Como o limite é 0, podemos concluir que:

$$
\log_2 n \in o(\sqrt{n})
$$

Ou seja, $ \log_2 n $ cresce mais lentamente que $ \sqrt{n} $, ou seja, $ \log_2 n $ está em uma classe de complexidade menor que $ \sqrt{n} $.

- Isso também implica:

$$
\log_2 n \in O(\sqrt{n}) \quad \text{e} \quad \sqrt{n} \in \Omega(\log_2 n)
$$

**(c) $ n! $ e $ 2^n $**

1. **Limite:**

$$
\lim_{n \to \infty} \frac{n!}{2^n} = \infty
$$

Como o limite é infinito, isso indica que $ n! $ cresce muito mais rapidamente que $ 2^n $.

2. **Conclusão em termos de Big-O, Theta e Omega:**

- Como o limite é infinito, podemos concluir que:

$$
n! \in \omega(2^n)
$$

Ou seja, $ n! $ cresce mais rapidamente que $ 2^n $, ou seja, $ n! $ está em uma classe de complexidade maior que $ 2^n $.

- Isso também implica:

$$
2^n \in O(n!) \quad \text{e} \quad n! \in \Omega(2^n)
$$

**Resumo Final:**

1. $ n(n - 1) $ e $ n^2 $ têm a mesma ordem de crescimento assintótica, ou seja:

$$
n(n - 1) \in \Theta(n^2)
$$

Além disso:

$$
n(n - 1) \in O(n^2) \quad \text{e} \quad n^2 \in \Omega(n(n - 1))
$$

2. $ \log_2 n $ cresce mais lentamente que $ \sqrt{n} $, ou seja:

$$
\log_2 n \in o(\sqrt{n})
$$

E também:

$$
\log_2 n \in O(\sqrt{n}) \quad \text{e} \quad \sqrt{n} \in \Omega(\log_2 n)
$$

3. $ n! $ cresce muito mais rapidamente que $ 2^n $, ou seja:

$$
n! \in \omega(2^n)
$$

E também:

$$
2^n \in O(n!) \quad \text{e} \quad n! \in \Omega(2^n)
$$

### Questão 9

Aqui está a revisão das equações com as modificações solicitadas, utilizando `$$` para as equações:

(a) $ \frac{n(n + 1)}{2} \in O(n^3) $

**Definição de Big-O:**
Uma função $ f(n) \in O(g(n)) $ se existem constantes positivas $ c $ e $ n_0 $ tais que:

$$
f(n) \leq c \cdot g(n) \quad \text{para todo } n \geq n_0
$$

Aqui, temos que $ f(n) = \frac{n(n + 1)}{2} $, que é uma função de ordem $ O(n^2) $, e $ g(n) = n^3 $. Sabemos que $ n^3 $ cresce mais rapidamente que $ n^2 $, então, existe uma constante $ c $ tal que:

$$
\frac{n(n + 1)}{2} \leq c \cdot n^3 \quad \text{para algum } n_0
$$

**Conclusão:** Esta afirmação é **verdadeira**. A função quadrática está, de fato, limitada por $ O(n^3) $.

(b) $ \frac{n(n + 1)}{2} \in \Theta(n^3) $

**Definição de Theta:**
Uma função $ f(n) \in \Theta(g(n)) $ se existem constantes positivas $ c_1 $, $ c_2 $ e $ n_0 $ tais que:

$$
c_1 \cdot g(n) \leq f(n) \leq c_2 \cdot g(n) \quad \text{para todo } n \geq n_0
$$

Aqui, $ f(n) = \frac{n(n + 1)}{2} $ e $ g(n) = n^3 $. Como $ f(n) $ é uma função quadrática e $ g(n) $ é cúbica, não podemos encontrar constantes $ c_1 $ e $ c_2 $ tais que $ f(n) $ esteja entre essas duas constantes multiplicadas por $ n^3 $ para $ n $ suficientemente grande.

**Conclusão:** Esta afirmação é **falsa**. A função quadrática não está em $ \Theta(n^3) $, pois a ordem de crescimento de $ n^3 $ é maior.

(c) $ \frac{n(n + 1)}{2} \in O(n^2) $

**Definição de Big-O:**
Novamente, temos $ f(n) = \frac{n(n + 1)}{2} $, que é uma função quadrática, e $ g(n) = n^2 $. Como $ f(n) $ é de ordem $ O(n^2) $, podemos facilmente ver que existe uma constante $ c $ tal que:

$$
\frac{n(n + 1)}{2} \leq c \cdot n^2 \quad \text{para algum } n_0
$$

**Conclusão:** Esta afirmação é **verdadeira**. A função $ \frac{n(n + 1)}{2} $ está em $ O(n^2) $, pois seu crescimento é quadrático.

(d) $ \frac{n(n + 1)}{2} \in \Omega(n) $

**Definição de Omega:**
Uma função $ f(n) \in \Omega(g(n)) $ se existem constantes positivas $ c $ e $ n_0 $ tais que:

$$
f(n) \geq c \cdot g(n) \quad \text{para todo } n \geq n_0
$$

Aqui, $ f(n) = \frac{n(n + 1)}{2} $, que é uma função quadrática, e $ g(n) = n $. Sabemos que $ \frac{n(n + 1)}{2} $ é uma função de ordem $ O(n^2) $, ou seja, cresce mais rápido que $ n $. Assim, existe uma constante $ c $ tal que:

$$
\frac{n(n + 1)}{2} \geq c \cdot n \quad \text{para todo } n \geq n_0
$$

**Conclusão:** Esta afirmação é **verdadeira**. A função quadrática está em $ \Omega(n) $, pois cresce mais rapidamente do que $ n $ para grandes valores de $ n $.

Resumo:

- **(a)** $ \frac{n(n + 1)}{2} \in O(n^3) $: **Verdadeiro**
- **(b)** $ \frac{n(n + 1)}{2} \in \Theta(n^3) $: **Falso**
- **(c)** $ \frac{n(n + 1)}{2} \in O(n^2) $: **Verdadeiro**
- **(d)** $ \frac{n(n + 1)}{2} \in \Omega(n) $: **Verdadeiro**

### Questão 10

$$\lim_{n \to \infty} \frac{p(n)}{n^k} = \lim_{n \to \infty} \frac{a_k n^k + a_{k-1} n^{k-1} + \ ... \ +a_0}{n^k} = \lim_{n \to \infty} a^k + \frac{a^{k-1}}{n} + \frac{a^{k-2}}{n^2} + \ ... \ + \frac{a_0}{n^k}$$

$$\therefore p(n) \in \Theta(n^k)$$

### Questão 11)

Para provar que funções exponenciais $ a^n $, para diferentes valores de $ a > 0 $, têm diferentes ordens de crescimento, vamos analisar o limite:

$$
\lim_{n \to \infty} \frac{a_1^n}{a_2^n}
$$

onde $ a_1 $ e $ a_2 $ são constantes positivas e $ a_1 \neq a_2 $.

Passo 1: Simplificação da expressão
Podemos reescrever a expressão da seguinte forma:

$$
\frac{a_1^n}{a_2^n} = \left( \frac{a_1}{a_2} \right)^n
$$

Isso mostra que o limite que precisamos analisar é:

$$
\lim_{n \to \infty} \left( \frac{a_1}{a_2} \right)^n
$$

Passo 2: Análise dos diferentes casos

Agora, vamos considerar os diferentes casos com base na relação entre $ a_1 $ e $ a_2 $.

Caso 1: $ a_1 > a_2 $

Se $ a_1 > a_2 $, então $ \frac{a_1}{a_2} > 1 $. Portanto, o termo $ \left( \frac{a_1}{a_2} \right)^n $ cresce exponencialmente à medida que $ n $ aumenta. Em termos de limite, temos:

$$
\lim_{n \to \infty} \left( \frac{a_1}{a_2} \right)^n = \infty
$$

Isso significa que $ a_1^n $ cresce mais rápido do que $ a_2^n $ quando $ a_1 > a_2 $.

Caso 2: $ a_1 < a_2 $

Se $ a_1 < a_2 $, então $ \frac{a_1}{a_2} < 1 $. Neste caso, o termo $ \left( \frac{a_1}{a_2} \right)^n $ decai para zero à medida que $ n $ aumenta. O limite é:

$$
\lim_{n \to \infty} \left( \frac{a_1}{a_2} \right)^n = 0
$$

Isso significa que $ a_1^n $ cresce mais lentamente que $ a_2^n $ quando $ a_1 < a_2 $.

Caso 3: $ a_1 = a_2 $

Se $ a_1 = a_2 $, então $ \frac{a_1}{a_2} = 1 $, e o limite se torna:

$$
\lim_{n \to \infty} 1^n = 1
$$

Nesse caso, $ a_1^n $ e $ a_2^n $ crescem na mesma taxa, e o comportamento das funções é idêntico.

Passo 3: Conclusão

A análise do limite $ \lim_{n \to \infty} \frac{a_1^n}{a_2^n} $ mostra que:

- Se $ a_1 > a_2 $, então $ a_1^n $ cresce mais rápido que $ a_2^n $.
- Se $ a_1 < a_2 $, então $ a_1^n $ cresce mais devagar que $ a_2^n $.
- Se $ a_1 = a_2 $, então $ a_1^n $ e $ a_2^n $ crescem na mesma taxa.

Isso prova que funções exponenciais com diferentes bases $ a_1 $ e $ a_2 $ têm ordens de crescimento diferentes, dependendo do valor de $ a_1 $ e $ a_2 $.

### Questão 12)

#### **Algoritmo 1: Mystery(n)**
```plaintext
ALGORITHM Mystery(n)
//Input: A nonnegative integer n
S ← 0
for i ← 1 to n do
    S ← S + i * i
return S
```

**(a) O que este algoritmo computa?**  
O algoritmo calcula a soma dos quadrados dos primeiros $ n $ números naturais, ou seja:  
$
S = 1^2 + 2^2 + 3^2 + \dots + n^2
$

**(b) Qual a operação básica deste algoritmo?**  
A operação básica é a multiplicação `i * i` (ou a adição `S + i * i`).

**(c) Quantas vezes esta operação básica é executada?**  
A operação é executada $ n $ vezes, uma vez para cada iteração do loop.

**(d) Qual a classe deste algoritmo em relação à eficiência?**  
O algoritmo tem complexidade de tempo $ O(n) $ (linear).

**(e) Sugira alguma melhora ou um novo algoritmo melhor:**  
Uma melhoria seria utilizar a fórmula matemática para a soma dos quadrados:  
$
S = \frac{n(n + 1)(2n + 1)}{6}
$  
Isso reduziria a complexidade para $ O(1) $ (constante), pois o cálculo seria direto sem loops.

---

#### **Algoritmo 2: Secret(A[0..n-1])**
```plaintext
ALGORITHM Secret(A[0..n-1])
//Input: An array A[0..n-1] of n real numbers
minval ← A[0]; maxval ← A[0]
for i ← 1 to n-1 do
    if A[i] < minval
        minval ← A[i]
    if A[i] > maxval
        maxval ← A[i]
return maxval - minval
```

**(a) O que este algoritmo computa?**  
O algoritmo calcula a amplitude (diferença entre o maior e o menor valor) de um array.

**(b) Qual a operação básica deste algoritmo?**  
A operação básica é a comparação `A[i] < minval` ou `A[i] > maxval`.

**(c) Quantas vezes esta operação básica é executada?**  
São feitas $ 2(n - 1) $ comparações (duas por elemento, exceto o primeiro).

**(d) Qual a classe deste algoritmo em relação à eficiência?**  
O algoritmo tem complexidade $ O(n) $ (linear), pois percorre o array uma vez.

**(e) Sugira alguma melhora ou um novo algoritmo melhor:**  
Não é possível melhorar a complexidade assintótica, pois qualquer algoritmo que precise verificar todos os elementos terá $ O(n) $.  
Uma pequena otimização seria reduzir o número de comparações por iteração, mas ainda seria $ O(n) $.

---

#### **Algoritmo 3: Enigma(A[0..n-1, 0..n-1])**
```plaintext
ALGORITHM Enigma(A[0..n-1, 0..n-1])
//Input: A matrix A[0..n-1, 0..n-1] of real numbers
for i ← 0 to n-2 do
    for j ← i+1 to n-1 do
        if A[i, j] ≠ A[j, i]
            return false
return true
```

**(a) O que este algoritmo computa?**  
O algoritmo verifica se uma matriz quadrada é simétrica (ou seja, $ A[i, j] = A[j, i] $ para todo $ i, j $).

**(b) Qual a operação básica deste algoritmo?**  
A operação básica é a comparação `A[i, j] ≠ A[j, i]`.

**(c) Quantas vezes esta operação básica é executada?**  
O número de comparações é:  
$
\sum_{i=0}^{n-2} (n - 1 - i) = \frac{n(n - 1)}{2}
$  
(metade dos elementos fora da diagonal principal).

**(d) Qual a classe deste algoritmo em relação à eficiência?**  
O algoritmo tem complexidade $ O(n^2) $ (quadrática), pois percorre metade da matriz.

**(e) Sugira alguma melhora ou um novo algoritmo melhor:**  
Não é possível reduzir a complexidade assintótica, pois todas as entradas relevantes devem ser verificadas.  
Uma pequena otimização seria interromper o loop assim que uma desigualdade for encontrada, mas no pior caso ainda seria $ O(n^2) $.

---

### **Resumo das Classes de Eficiência:**
| Algoritmo | Operação Básica       | Nº de Execuções | Complexidade |
|-----------|-----------------------|-----------------|--------------|
| Mystery   | Multiplicação (`i*i`) | $ n $         | $ O(n) $   |
| Secret    | Comparação (`A[i] <>`) | $ 2(n-1) $    | $ O(n) $   |
| Enigma    | Comparação (`A[i,j]`) | $ \frac{n(n-1)}{2} $ | $ O(n^2) $ |

### Questão 14)

In [ ]:
def alg(n):
    # Caso base
    if n == 0:
        return 1
    # Passo recursivo
    return 2 * alg(n - 1)

# Testando o algoritmo
n = 3
resultado = alg(n)
print(f"2^{n} = {resultado}")

2^3 = 8


### Questão 15)

#### **Algoritmo: Riddle(A[0..n-1])**
```plaintext
ALGORITHM Riddle(A[0..n-1])
//Input: An array A[0..n-1] of real numbers
if n = 1
    return A[0]
else
    temp ← Riddle(A[0..n-2])
    if temp ≤ A[n-1]
        return temp
    else
        return A[n-1]
```

---

#### **(a) O que este algoritmo faz?**
O algoritmo **encontra o menor elemento** em um array de números reais.  
- **Caso base (`n = 1`)**: Retorna o único elemento do array.  
- **Passo recursivo**: Compara o menor elemento do subarray `A[0..n-2]` (calculado recursivamente) com o último elemento `A[n-1]` e retorna o menor entre os dois.

**Exemplo**:  
Para `A = [3, 1, 4, 2]`:  
- Chama `Riddle([3, 1, 4])`, que retorna `1` (menor entre `3`, `1`, `4`).  
- Compara `1` (temp) com `2` (último elemento) e retorna `1`.

---

#### **(b) Relação de recorrência e solução**

**(i) Operação básica**:  
A comparação `temp ≤ A[n-1]` (ou a atribuição de `temp`).  

**(ii) Relação de recorrência**:  
Seja \( C(n) \) o número de operações básicas para um array de tamanho \( n \):  
- **Caso base**: \( C(1) = 0 \) (nenhuma comparação).  
- **Passo recursivo**: \( C(n) = C(n-1) + 1 \), onde:  
  - \( C(n-1) \): Chamada recursiva para `A[0..n-2]`.  
  - \( +1 \): Comparação entre `temp` e `A[n-1]`.  

**(iii) Resolução da recorrência**:  
A recorrência é linear e homogênea. Podemos expandir:  
\[
C(n) = C(n-1) + 1 = C(n-2) + 1 + 1 = \dots = C(1) + (n-1) = 0 + (n-1) = n-1.
\]  
Portanto, \( C(n) = n - 1 \).

**(iv) Complexidade**:  
O algoritmo executa \( n-1 \) comparações no total, logo sua complexidade é \( O(n) \).

---

#### **(c) Observações**
- **Eficiência**: O algoritmo é eficiente para encontrar o mínimo em \( O(n) \), que é ótimo para o problema (todo algoritmo deve verificar todos os elementos no pior caso).  
- **Melhorias**:  
  - **Versão iterativa**: Pode ser reescrito sem recursão (evitando overhead de chamadas), mas mantendo \( O(n) \).  
  - **Paralelização**: Em ambientes paralelos, divisão do array em subarrays pode acelerar, mas a complexidade assintótica permanece \( O(n) \).

---

### **Resumo Final**
| Item               | Resposta                                                                 |
|--------------------|--------------------------------------------------------------------------|
| **Função**         | Encontra o menor elemento em um array.                                  |
| **Operação básica**| Comparação `temp ≤ A[n-1]`.                                             |
| **Relação**        | \( C(n) = C(n-1) + 1 \), com \( C(1) = 0 \).                           |
| **Solução**        | \( C(n) = n - 1 \).                                                    |
| **Complexidade**   | \( O(n) \) (linear).                                                   |

### Questão 16)

#### **Algoritmo: GraphComplete(A[0..n-1, 0..n-1])**
```plaintext
ALGORITHM GraphComplete(A[0..n-1, 0..n-1])
//Input: Adjacency matrix A[0..n-1, 0..n-1] of an undirected graph G
//Output: 1 (true) if G is complete, 0 (false) otherwise
if n = 1
    return 1  // Um grafo com 1 vértice é completo por definição
else
    if not GraphComplete(A[0..n-2, 0..n-2])
        return 0
    else
        for j ← 0 to n-2 do
            if A[n-1, j] = 0
                return 0
        return 1
```

---

#### **(a) O que o algoritmo faz?**
O algoritmo verifica se um grafo não direcionado (representado por sua **matriz de adjacência**) é **completo**.  
- Um grafo completo tem uma aresta entre todo par de vértices distintos (ou seja, \( A[i, j] = 1 \) para todo \( i \neq j \)).  
- O algoritmo usa **recursão** para verificar a completude:  
  1. **Caso base**: Grafo com 1 vértice é trivialmente completo.  
  2. **Passo recursivo**:  
     - Verifica se o subgrafo \( A[0..n-2, 0..n-2] \) é completo.  
     - Se for, verifica se o vértice \( n-1 \) está conectado a todos os outros vértices \( 0 \) a \( n-2 \).

---

#### **(b) Operação básica**  
A operação básica é a **comparação** `A[n-1, j] = 0` (verificação de ausência de aresta).

---

#### **(c) Número de operações básicas**  
Seja \( T(n) \) o número de operações para uma matriz \( n \times n \):  
- **Caso base**: \( T(1) = 0 \) (nenhuma comparação).  
- **Passo recursivo**:  
  - Chamada recursiva para \( T(n-1) \).  
  - Loop para verificar arestas do vértice \( n-1 \): \( n-1 \) comparações.  

**Relação de recorrência**:  
\[
T(n) = T(n-1) + (n-1), \quad T(1) = 0
\]

**Resolução**:  
Expandindo a recorrência:  
\[
T(n) = T(n-1) + (n-1) = T(n-2) + (n-2) + (n-1) = \dots = \sum_{k=1}^{n-1} k = \frac{n(n-1)}{2}.
\]  
Portanto, \( T(n) = \frac{n(n-1)}{2} \).

---

#### **(d) Classe de eficiência no pior caso**  
- **Número total de comparações**: \( \frac{n(n-1)}{2} \) (equivalente a verificar todas as entradas acima da diagonal principal).  
- **Complexidade**: \( O(n^2) \) (quadrática), pois:  
  \[
  \frac{n(n-1)}{2} \in \Theta(n^2).
  \]

---

#### **(e) Melhorias possíveis**  
1. **Abordagem iterativa**:  
   - Evitar overhead de recursão usando loops aninhados para verificar \( A[i, j] = 1 \) para \( i \neq j \).  
   - Complexidade permanece \( O(n^2) \), mas reduz custos constantes.  

2. **Paralelização**:  
   - Verificar arestas em paralelo (em hardware apropriado), mas a complexidade assintótica não muda.  

3. **Otimização para grafos esparsos**:  
   - Se o grafo é representado por lista de adjacências, pode-se interromper a verificação assim que um vértice com grau \( \neq n-1 \) for encontrado.  
   - **No entanto**, para matriz de adjacência, a verificação \( O(n^2) \) é inevitável no pior caso.

---

### **Resumo Final**
| Item               | Resposta                                                                 |
|--------------------|--------------------------------------------------------------------------|
| **Função**         | Verifica se um grafo não direcionado é completo.                        |
| **Operação básica**| Comparação `A[i, j] = 0`.                                              |
| **Nº de operações**| \( \frac{n(n-1)}{2} \) (triangular).                                   |
| **Complexidade**   | \( O(n^2) \) (quadrática).                                             |
| **Melhorias**      | Iteração sem recursão, mas mantendo \( O(n^2) \).                       |

**Conclusão**: O algoritmo é ótimo para matriz de adjacência, pois qualquer solução exige verificar \( \Omega(n^2) \) entradas no pior caso.

# Lista 2

## Chapter 2 - Apostila

### Questão 3)

In [ ]:
# Alternativa a
# 1. Usamos list comprehension para gerar uma lista de números primos no intervalo de 2 a 11.
# 2. A condição all(x % d != 0 for d in range(2, int(x ** 0.5) + 1)) verifica se um número x é primo:
#   - Um número é primo se não for divisível por nenhum outro número além de 1 e ele mesmo.
#   - Para eficiência, testamos apenas divisores até int(x ** 0.5) + 1, pois se x tiver um fator maior que sua raiz quadrada,
#     ele terá um fator menor também.
# 3. O range range(2, 12) foi escolhido para incluir os cinco primeiros números primos [2, 3, 5, 7, 11].
fst_primes = [x for x in range(2, 12)
              if all(x % d != 0 for d in range(2, int(x ** 0.5) + 1))]

print(fst_primes)

print()

# Alternativa b
def remove_even_numbers(array):
  return [array[i] for i in range(len(array)) if array[i] % 2 != 0]

print(remove_even_numbers([1, 2, 3, 4, 5, 6]))

print()

# Alternativa c
def find_second_largest(array):
    largest = second_largest = float('-inf')  # Inicializamos com um valor muito pequeno

    for num in array:
        if num > largest:
            second_largest = largest  # Atualiza o segundo maior número
            largest = num  # Atualiza o maior número
        elif num > second_largest and num != largest:
            second_largest = num  # Atualiza apenas o segundo maior número

    if second_largest == float('-inf'):
        raise ValueError("Todos os elementos do array são iguais.")

    return second_largest

print(find_second_largest([10, 20, 4, 45, 99]))

print()

# Alternativa d
def flatten_nested_list(array):
    return [item for sublist in array for item in sublist]

print(flatten_nested_list([[1, 2], [3, 4], [5]]))

print()

# Alternativa e
def rotate_list(array, pos):
    return [array[(i - pos) % len(array)] for i in range(len(array))]

print(rotate_list([1, 2, 3, 4, 5], 2))

print()

# Alternativa f
def count_occurrences(array):
    counts = {}
    for item in array:
        counts[item] = counts.get(item, 0) + 1 # Se 'item' já existe no dicionário 'counts', incrementa seu valor em 1.
                                              # Caso contrário, 'get(item, 0)' retorna 0 (valor padrão), e o item é adicionado com valor 1.
    return counts

print(count_occurrences([1, 1, 2, 3, 3, 3, 4]))

[2, 3, 5, 7, 11]

[1, 3, 5]

45

[1, 2, 3, 4, 5]

[4, 5, 1, 2, 3]

{1: 2, 2: 1, 3: 3, 4: 1}


### Questão 4

In [ ]:
# Alternativa a
def convert_list_of_tuples_to_dict(array_of_tuples):
  return dict(array_of_tuples)

print(convert_list_of_tuples_to_dict([(1,'a'), (2, 'b'), (3, 'c')]))

print()

# Alternativa b
def swap_fst_and_lst_element(tup):
  return (tup[-1], *tup[1:-1], tup[0])

print(swap_fst_and_lst_element((1, 2, 3, 4, 5)))

print()

# Alternativa c
def find_minmax_tuple(tup):
    min_value, max_value = tup[0], tup[0]  # Inicializa com o primeiro elemento

    for num in tup:  # Percorre a tupla
        if num < min_value:
            min_value = num  # Atualiza o menor valor
        if num > max_value:
            max_value = num  # Atualiza o maior valor

    return (max_value, min_value)

print(find_minmax_tuple((4, 7, 1, 9)))

print()

# Alternativa d
def concatenate_tuple_into_str(tup):
  string = ''
  for char in tup:
    string += str(char)
  return string

print(concatenate_tuple_into_str((1, 2, 3, 4)))

{1: 'a', 2: 'b', 3: 'c'}

(5, 2, 3, 4, 1)

(9, 1)

1234


### Questão 5

In [ ]:
# Alternativa a
def merge_dicts_sum_values(dict1, dict2):
    # Merge dictionaries, summing values of common keys
    merged_dict = dict1.copy()  # Start with a copy of dict1 to avoid modifying it
    for key, value in dict2.items():
        if key in merged_dict:
            merged_dict[key] += value  # Sum the values for common keys
        else:
            merged_dict[key] = value  # Add the key-value pair if the key is not in dict1
    return merged_dict

print(merge_dicts_sum_values({'a' : 1, 'b' : 2}, {'b' : 3, 'c' : 4}))

print()

# Alternativa b
def invert_dict(d):
    return {value: key for key, value in d.items()}

print(invert_dict({'a' : 1, 'b' : 2}))

print()

# Alternativa c
def most_frequent_value(d):
    # Cria um dicionário para armazenar a contagem de cada valor
    frequency = {}

    # Conta as ocorrências de cada valor no dicionário
    for value in d.values():
        frequency[value] = frequency.get(value, 0) + 1

    # Encontra o valor com a maior contagem
    most_frequent = max(frequency, key=frequency.get)

    return most_frequent

print(most_frequent_value({'a': 3, 'b': 2, 'c': 3}))

print()

# Alternativa d
def group_words_by_first_letter(words):
    # Cria um dicionário vazio para armazenar os grupos de palavras
    grouped_words = {}

    # Itera sobre a lista de palavras
    for word in words:
        # Obtém a primeira letra da palavra
        first_letter = word[0].lower()  # Usar .lower() para garantir que a letra seja minúscula

        # Adiciona a palavra ao grupo correspondente à primeira letra
        if first_letter not in grouped_words:
            grouped_words[first_letter] = [word]  # Se a letra não existir, cria uma nova lista
        else:
            grouped_words[first_letter].append(word)  # Se a letra existir, adiciona a palavra à lista

    return grouped_words

print(group_words_by_first_letter(["apple", "banana", "apricot", "blueberry", "cherry"]))

print()

# Alternativa e
def find_key_with_max_value(d):
    # Inicializa o primeiro item do dicionário como o máximo
    max_key = max(d, key=d.get)

    return max_key

print(find_key_with_max_value({'a': 10, 'b': 25, 'c': 17}))

{'a': 1, 'b': 5, 'c': 4}

{1: 'a', 2: 'b'}

3

{'a': ['apple', 'apricot'], 'b': ['banana', 'blueberry'], 'c': ['cherry']}

b


### Questão 6

In [ ]:
# Alternativa a
def set_operations(set1, set2):
    union_set = set1 | set2  # União dos conjuntos
    intersection_set = set1 & set2  # Interseção dos conjuntos
    difference_set1 = set1 - set2  # Diferença (set1 - set2)
    difference_set2 = set2 - set1  # Diferença (set2 - set1)

    return {
        "Union": union_set,
        "Intersection": intersection_set,
        "Difference (Set 1 - Set 2)": difference_set1,
        "Difference (Set 2 - Set 1)": difference_set2
    }

set1 = {1, 2, 3}
set2 = {3, 4, 5}
output = set_operations(set1, set2)

for key, value in output.items():
    print(f"{key}: {value}")

print()

# Alternativa b
def symmetric_difference(set1, set2):
    return set1 ^ set2  # Operador XOR para diferença simétrica

print(symmetric_difference(set([1, 2, 3]), set([2, 3, 4])))

print()

# Alternativa c
def are_disjoint(set1, set2):
    return set1.isdisjoint(set2)  # Retorna True se não houver elementos em comum

print(are_disjoint(set([1, 2, 3]), set([4, 5, 6])))
print(are_disjoint(set([1, 2, 3]), set([3, 4, 5])))

Union: {1, 2, 3, 4, 5}
Intersection: {3}
Difference (Set 1 - Set 2): {1, 2}
Difference (Set 2 - Set 1): {4, 5}

{1, 4}

True
False


## Chapter 3 - Apostila

### Questão 1

In [ ]:
class Rectangle:
  def __init__(self, length, width):
    self.length = length
    self.width = width

  def area(self):
    self.area = self.length * self.width
    return self.area

  def perimeter(self):
    self.perimeter = 2 * (self.length + self.width)
    return self.perimeter

rectangle = Rectangle(4, 5)
print(rectangle.area())
print(rectangle.perimeter())

20
18


### Questão 2

In [ ]:
import math

class Circle:
  def __init__(self, radius):
    self.radius = radius

  def circumference(self):
    self.circumference = 2 * math.pi * self.radius
    return self.circumference

  def area(self):
    self.area = math.pi * self.radius ** 2
    return self.area

circle = Circle(5)
print(circle.area())

78.53981633974483


### Questão 3

In [ ]:
class Student:
  def __init__(self, name, age, grades):
    self.name = name
    self.age = age
    self.grades = grades

  def average(self):
    self.avg = sum(self.grades) / len(self.grades)
    return self.avg

  def is_passing(self):
    return self.average() >= 60

alice = Student("Alice", 20, [80, 90])
print(alice.average())
print(alice.is_passing())

85.0
True


### Questão 4

In [ ]:
class BankAccount:
  def __init__(self, balance):
    self.balance = balance

  def deposit(self, amount):
    self.balance += amount

  def withdraw(self, amount):
    if self.balance >= amount:
      self.balance -= amount
    else:
      print("Insufficient funds")

acc = BankAccount(100)
acc.deposit(50)
print(acc.balance)

150


### Questão 5

In [ ]:
# Superclasse Vehicle
class Vehicle:
    def __init__(self, make, model):
        self.make = make  # Marca do veículo
        self.model = model  # Modelo do veículo

    def display_info(self):
        return f"Make: {self.make}, Model: {self.model}"

# Subclasse Car (herda de Vehicle)
class Car(Vehicle):
    def __init__(self, make, model, doors):
        super().__init__(make, model)  # Chama o construtor da superclasse
        self.doors = doors  # Número de portas

    def display_info(self):
        return f"{super().display_info()}, Doors: {self.doors}"

# Subclasse Bike (herda de Vehicle)
class Bike(Vehicle):
    def __init__(self, make, model, bike_type):
        super().__init__(make, model)  # Chama o construtor da superclasse
        self.bike_type = bike_type  # Tipo de bicicleta (ex: "mountain", "road")

    def display_info(self):
        return f"{super().display_info()}, Type: {self.bike_type}"

# Teste da implementação
car = Car("Ford", "Mustang", 4)
bike = Bike("Giant", "Escape 3", "road")

print(car.display_info())  # Output: Make: Ford, Model: Mustang, Doors: 4
print(bike.display_info())  # Output: Make: Giant, Model: Escape 3, Type: road

Make: Ford, Model: Mustang, Doors: 4
Make: Giant, Model: Escape 3, Type: road


## Chapter 5 - Apostila

### Questão 1

In [ ]:
import timeit

def sum_using_loop(k):
    """Soma os primeiros k números naturais usando um loop."""
    total = 0
    for i in range(1, k + 1):
        total += i
    return total

def sum_using_formula(k):
    """Soma os primeiros k números naturais usando a fórmula S = k(k+1)/2."""
    return k * (k + 1) // 2  # Fórmula fechada para a soma

# Valor de k para teste
k = 1000

# Medindo tempo com timeit para maior precisão
loop_time = timeit.timeit(lambda: sum_using_loop(k), number=1000) / 1000
formula_time = timeit.timeit(lambda: sum_using_formula(k), number=1000) / 1000

# Exibindo os resultados em milissegundos (ms)
print(f"Time for loop: {loop_time * 1e3:.6f} ms")
print(f"Time for formula: {formula_time * 1e3:.6f} ms")

Time for loop: 0.094332 ms
Time for formula: 0.000338 ms


#### **Sumário de Complexidade**

A análise da complexidade de algoritmos ajuda a entender o **desempenho** e a **eficiência** de diferentes abordagens. Aqui estão os conceitos-chave aplicados ao nosso caso:

---

### **1. Complexidade do Loop**
A função `sum_using_loop(k)` percorre todos os números de `1` a `k` e realiza `k` somas.

- **Número de operações**: Aproximadamente $k$ somas.
- **Complexidade:**  
  $$ O(k) $$
- **Tempo de execução (para $k=1000$):** `0.094332 ms`  

Isso significa que, à medida que $k$ cresce, o tempo de execução **aumenta linearmente**.

---

### **2. Complexidade da Fórmula**
A função `sum_using_formula(k)` calcula a soma diretamente com a fórmula matemática:

$$ S = \frac{k(k+1)}{2} $$

- **Número de operações:** Apenas **3** operações (multiplicação, adição e divisão inteira).
- **Complexidade:**  
  $$ O(1) $$
- **Tempo de execução (para $k=1000$):** `0.000338 ms`  

Essa abordagem **não depende** do valor de $k$, pois sempre executa a mesma quantidade de operações, tornando-se **extremamente eficiente** para valores grandes.

---

### **3. Comparação e Implicações**
| Método | Complexidade | Tempo para $k=1000$ |
|--------|------------|-----------------|
| Loop ($O(k)$) | Linear | `0.094332 ms` |
| Fórmula ($O(1)$) | Constante | `0.000338 ms` |

- **O loop cresce proporcionalmente a $k$**, tornando-se lento para valores grandes.
- **A fórmula é instantânea**, pois **independe de $k$**, sendo mais escalável.

Se $k$ aumentar para **1 milhão**, a fórmula continuará extremamente rápida, enquanto o loop levará muito mais tempo!

🚀 **Conclusão:** A abordagem com **fórmula fechada** é **muito mais eficiente** para somas grandes, devido à sua **complexidade constante $O(1)$**.

### Questão 2

#### **Analysis of Time Complexity for Common List Operations in Python**

Python lists are implemented as **dynamic arrays**, which impacts the complexity of various operations. Below, we analyze the time complexity of common list operations, including **appending, indexing, and slicing**, with examples and explanations.

---

## **1. Appending an Element (`list.append()`)**
### **Example:**
```python
lst = [1, 2, 3]
lst.append(4)  # Adding an element at the end
print(lst)  # Output: [1, 2, 3, 4]
```
### **Complexity Analysis:**
- Appending an element to the end of the list is **$O(1)$ (amortized)**.
- In most cases, Python lists have extra allocated space, so adding an element does not require resizing.
- If resizing is needed (i.e., when capacity is exceeded), a **new larger array** is allocated, and all elements are copied, making it **$O(n)$ in worst case**.
- However, since resizing happens **infrequently**, the **average complexity remains $O(1)$**.

---

## **2. Indexing an Element (`list[i]`)**
### **Example:**
```python
lst = [10, 20, 30, 40]
print(lst[2])  # Output: 30
```
### **Complexity Analysis:**
- Accessing an element by index takes **constant time $O(1)$**.
- Lists are stored as contiguous memory arrays, allowing direct access via an index lookup.

---

## **3. Slicing a List (`list[start:stop:step]`)**
### **Example:**
```python
lst = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
slice_lst = lst[2:8:2]  # [2, 4, 6]
print(slice_lst)
```
### **Complexity Analysis:**
- **Slicing creates a new list**, which requires copying elements.
- The number of elements copied is **proportional to the size of the slice**.
- If the slice contains **$m$ elements**, the complexity is **$O(m)$**.

---

## **4. Inserting an Element at a Specific Index (`list.insert()`)**
### **Example:**
```python
lst = [1, 2, 4, 5]
lst.insert(2, 3)  # Inserts 3 at index 2
print(lst)  # Output: [1, 2, 3, 4, 5]
```
### **Complexity Analysis:**
- Inserting an element at the **beginning or middle** shifts all subsequent elements.
- In the worst case (inserting at the start), all **$n$ elements** are shifted right.
- **Time Complexity: $O(n)$**.

---

## **5. Deleting an Element (`del list[i]` or `list.remove(x)`)**
### **Example:**
```python
lst = [1, 2, 3, 4, 5]
del lst[2]  # Removes element at index 2
print(lst)  # Output: [1, 2, 4, 5]
```
### **Complexity Analysis:**
- Deleting an element from a list **shifts all elements after it**.
- **Time Complexity: $O(n)$** in the worst case (deleting the first element).

---

## **6. Finding an Element (`list.index(x)`)**
### **Example:**
```python
lst = [10, 20, 30, 40, 50]
index = lst.index(30)  # Output: 2
```
### **Complexity Analysis:**
- `list.index(x)` **searches through the list**.
- **Worst case: $O(n)$** if the element is at the end or absent.

---

## **Summary of Complexities**
| Operation        | Complexity |
|-----------------|------------|
| `append(x)`     | $O(1)$ (amortized) |
| `list[i]`       | $O(1)$ |
| `list[a:b:c]`   | $O(m)$ (size of slice) |
| `insert(i, x)`  | $O(n)$ |
| `del list[i]`   | $O(n)$ |
| `list.index(x)` | $O(n)$ |

### **Conclusion**
- **Appending and indexing are very efficient** ($O(1)$).
- **Slicing and deleting elements require copying** ($O(n)$ in worst case).
- **Insertion and searching operations degrade to $O(n)$ for large lists**.

🚀 **Optimizations:** If frequent insertions/removals are needed, **consider using a `collections.deque`**, which provides **$O(1)$** insert/delete operations at both ends!

### Questão 3

In [ ]:
import timeit
import random

# Gerar um dicionário com n elementos
n = 10**6  # 1 milhão de elementos
test_dict = {i: i*2 for i in range(n)}

# Selecionar uma chave aleatória para teste de acesso
key = random.randint(0, n - 1)
new_key = n  # A chave que vamos adicionar para remoção

# Inserir a chave no dicionário para garantir que ela existe antes de removê-la
test_dict[new_key] = new_key * 2

# Medir tempo de acesso
access_time = timeit.timeit(lambda: test_dict[key], number=100000) / 100000 * 1000  # em ms

# Medir tempo de inserção (já garantimos que a chave foi inserida)
insert_time = timeit.timeit(lambda: test_dict.__setitem__(new_key, new_key * 2), number=100000) / 100000 * 1000

# Medir tempo de remoção (usando pop, garantindo que a chave existe)
delete_time = timeit.timeit(lambda: test_dict.pop(new_key, None), number=100000) / 100000 * 1000

# Exibir resultados
print(f"Access Time: {access_time:.6f} ms")
print(f"Insertion Time: {insert_time:.6f} ms")
print(f"Deletion Time: {delete_time:.6f} ms")

Access Time: 0.000078 ms
Insertion Time: 0.000173 ms
Deletion Time: 0.000087 ms


#### **Análise dos Resultados**

Com base nos tempos de execução medidos para as operações de acesso, inserção e remoção em um dicionário Python, podemos comparar os resultados obtidos com as complexidades teóricas esperadas.

### **1. Acesso ao Dicionário**

**Tempo de Acesso:**  
- **Resultado Medido:** `0.000078 ms`
- **Complexidade Teórica:**  
  O acesso a um dicionário pelo **índice** de uma chave, como `test_dict[key]`, tem uma **complexidade média de tempo O(1)** devido à implementação interna de tabelas de hash.
  
**Conclusão:**  
- O tempo de execução do acesso está de acordo com a **complexidade constante O(1)**. O resultado é muito baixo, refletindo a eficiência da operação.

---

### **2. Inserção no Dicionário**

**Tempo de Inserção:**  
- **Resultado Medido:** `0.000173 ms`
- **Complexidade Teórica:**  
  A inserção de um item em um dicionário, ou seja, a operação `test_dict[key] = value`, tem **complexidade O(1)** em média, pois envolve a computação do hash da chave e a inserção no slot correspondente na tabela de hash.

**Conclusão:**  
- O tempo de execução de inserção também está dentro da faixa esperada para uma **complexidade constante O(1)**, embora a inserção de novos elementos possa, em algumas situações, envolver uma **remoção de colisões** ou **realocação** de memória, mas isso ocorre de maneira rara e não afeta a complexidade média.

---

### **3. Remoção do Dicionário**

**Tempo de Remoção:**  
- **Resultado Medido:** `0.000087 ms`
- **Complexidade Teórica:**  
  A remoção de um item de um dicionário, como `test_dict.pop(key)`, também tem **complexidade O(1)** em média. A operação envolve encontrar a chave através de seu hash, removê-la e ajustar a tabela de hash se necessário.

**Conclusão:**  
- O tempo de execução da remoção segue a **complexidade constante O(1)**, como esperado. Assim como a inserção, a operação de remoção é geralmente eficiente devido à tabela de hash.

---

### **Comparação com a Teoria**

A análise dos resultados confirma as **complexidades esperadas**:

- **Acesso:** O tempo de `0.000078 ms` está em linha com a **complexidade O(1)**, já que o dicionário em Python utiliza uma tabela de hash para acessos rápidos.
- **Inserção:** A operação de inserção leva `0.000173 ms`, que também se alinha com a **complexidade O(1)**.
- **Remoção:** O tempo de remoção foi de `0.000087 ms`, novamente refletindo uma **complexidade O(1)**.

### **Conclusão Final**

As operações de **acesso**, **inserção** e **remoção** em um dicionário Python têm uma **complexidade média de O(1)**. O tempo de execução observado para essas operações é muito baixo, o que é típico para dicionários implementados com tabelas de hash. A análise experimental confirma que, em termos de complexidade teórica e prática, essas operações são extremamente eficientes.

Se você tiver mais dúvidas ou precisar de mais análises, sinta-se à vontade para perguntar!

### Questão 4

#### **Análise de Complexidade de Tempo das Funções**

A seguir, realizaremos a análise de complexidade de tempo no pior caso (worst-case) para cada função das alternativas fornecidas.

#### **Alternativa a: Lista de Números Primos**
```python
fst_primes = [x for x in range(2, 12) if all(x % d != 0 for d in range(2, int(x ** 0.5) + 1))]
```
- **Análise:** Para cada número `x` no intervalo de 2 a 11, a função `all` verifica se `x` é divisível por algum número até a raiz quadrada de `x`. O número de divisores a ser verificado cresce aproximadamente como a raiz quadrada de `x`.
- **Complexidade no pior caso:** A iteração principal percorre os números de 2 a 11 (9 números), e a verificação de primalidade para cada número envolve iterar até aproximadamente a raiz quadrada de `x`. A complexidade total é \(O(n \cdot \sqrt{m})\), onde `n` é o número de elementos na lista (9 elementos) e `m` é o valor máximo no intervalo (11).
- **Pior caso:** \(O(n \cdot \sqrt{m}) = O(9 \cdot \sqrt{11}) = O(9 \cdot 3.32) = O(30)\).

#### **Alternativa b: Remover Números Pares**
```python
def remove_even_numbers(array):
    return [array[i] for i in range(len(array)) if array[i] % 2 != 0]
```
- **Análise:** A função percorre toda a lista, verificando para cada item se é ímpar e adicionando à nova lista.
- **Complexidade no pior caso:** A função percorre todos os `n` elementos da lista e realiza uma operação de módulo constante \(O(1)\) para cada elemento.
- **Pior caso:** \(O(n)\), onde `n` é o número de elementos na lista.

#### **Alternativa c: Encontrar o Segundo Maior**
```python
def find_second_largest(array):
    largest = second_largest = float('-inf')
    for num in array:
        if num > largest:
            second_largest = largest
            largest = num
        elif num > second_largest and num != largest:
            second_largest = num
    if second_largest == float('-inf'):
        raise ValueError("Todos os elementos do array são iguais.")
    return second_largest
```
- **Análise:** A função percorre a lista uma vez, comparando cada elemento para atualizar o maior e o segundo maior valor.
- **Complexidade no pior caso:** A função percorre todos os `n` elementos da lista, realizando comparações constantes \(O(1)\) para cada elemento.
- **Pior caso:** \(O(n)\), onde `n` é o número de elementos na lista.

#### **Alternativa d: Achatar Lista Aninhada**
```python
def flatten_nested_list(array):
    return [item for sublist in array for item in sublist]
```
- **Análise:** A função percorre todas as sublistas e todos os itens dentro de cada sublista.
- **Complexidade no pior caso:** A função percorre todos os `n` itens nas sublistas e, para cada sublista, percorre todos os seus elementos. Se o número total de elementos for `m`, a complexidade é \(O(m)\).
- **Pior caso:** \(O(m)\), onde `m` é o número total de itens nas sublistas.

#### **Alternativa e: Rotacionar Lista**
```python
def rotate_list(array, pos):
    return [array[(i - pos) % len(array)] for i in range(len(array))]
```
- **Análise:** A função percorre toda a lista e calcula um novo índice para cada elemento. A operação de módulo é \(O(1)\).
- **Complexidade no pior caso:** A função percorre todos os `n` elementos da lista, fazendo uma operação constante para cada.
- **Pior caso:** \(O(n)\), onde `n` é o número de elementos na lista.

#### **Alternativa f: Contar Ocorrências**
```python
def count_occurrences(array):
    counts = {}
    for item in array:
        counts[item] = counts.get(item, 0) + 1
    return counts
```
- **Análise:** A função percorre todos os `n` itens da lista, fazendo uma operação constante \(O(1)\) de verificação e atualização no dicionário.
- **Complexidade no pior caso:** \(O(n)\), onde `n` é o número de elementos na lista.

---

### **Funções Trabalhando com Tuplas**

#### **Alternativa a: Converter Lista de Tuplas para Dicionário**
```python
def convert_list_of_tuples_to_dict(array_of_tuples):
    return dict(array_of_tuples)
```
- **Análise:** A função percorre todos os elementos da lista de tuplas e os adiciona a um dicionário.
- **Complexidade no pior caso:** A função percorre todos os `n` elementos da lista, com cada operação de inserção no dicionário sendo \(O(1)\).
- **Pior caso:** \(O(n)\), onde `n` é o número de tuplas.

#### **Alternativa b: Trocar Primeiro e Último Elemento da Tupla**
```python
def swap_fst_and_lst_element(tup):
    return (tup[-1], *tup[1:-1], tup[0])
```
- **Análise:** A função realiza operações constantes para acessar os elementos e rearranjar os itens da tupla.
- **Complexidade no pior caso:** \(O(n)\), onde `n` é o número de elementos na tupla (porque a tupla precisa ser convertida em uma nova tupla, e isso envolve percorrer todos os elementos).

#### **Alternativa c: Encontrar Mínimo e Máximo da Tupla**
```python
def find_minmax_tuple(tup):
    min_value, max_value = tup[0], tup[0]
    for num in tup:
        if num < min_value:
            min_value = num
        if num > max_value:
            max_value = num
    return (max_value, min_value)
```
- **Análise:** A função percorre todos os elementos da tupla para encontrar o mínimo e o máximo.
- **Complexidade no pior caso:** \(O(n)\), onde `n` é o número de elementos na tupla.

#### **Alternativa d: Concatenar Tupla em uma String**
```python
def concatenate_tuple_into_str(tup):
    string = ''
    for char in tup:
        string += str(char)
    return string
```
- **Análise:** A função percorre todos os elementos da tupla e os converte para string, concatenando-os à variável `string`. Como strings são imutáveis, cada concatenação pode ser \(O(n)\) no pior caso.
- **Complexidade no pior caso:** \(O(n^2)\), onde `n` é o número de elementos na tupla, devido à concatenação repetida de strings.

---

### **Funções Trabalhando com Dicionários**

#### **Alternativa a: Mesclar Dicionários Somando os Valores**
```python
def merge_dicts_sum_values(dict1, dict2):
    merged_dict = dict1.copy()
    for key, value in dict2.items():
        if key in merged_dict:
            merged_dict[key] += value
        else:
            merged_dict[key] = value
    return merged_dict
```
- **Análise:** A função percorre todos os itens em `dict2`, e para cada item, verifica se a chave já existe em `dict1`. A operação de verificação e inserção no dicionário é \(O(1)\).
- **Complexidade no pior caso:** \(O(n + m)\), onde `n` e `m` são os tamanhos de `dict1` e `dict2`, respectivamente.

#### **Alternativa b: Inverter Dicionário**
```python
def invert_dict(d):
    return {value: key for key, value in d.items()}
```
- **Análise:** A função percorre todos os itens no dicionário `d` e cria um novo dicionário.
- **Complexidade no pior caso:** \(O(n)\), onde `n` é o número de itens no dicionário.

#### **Alternativa c: Valor Mais Frequente no Dicionário**
```python
def most_frequent_value(d):
    frequency = {}
    for value in d.values():
        frequency[value] = frequency.get(value, 0) + 1
    most_frequent = max(frequency, key=frequency.get)
    return most_frequent
```
- **Análise:** A função percorre todos os valores no dicionário e conta a frequência de cada valor. Em seguida, encontra o valor mais frequente.
- **Complexidade no pior caso:** \(O(n)\), onde `n` é o número de itens no dicionário.

#### **Alternativa d: Agrupar Palavras pela Primeira Letra**
```python
def group_words_by_first_letter(words):
    grouped_words = {}
    for word in words:
        first_letter = word[0].lower()
        if first_letter not in grouped_words:
            grouped_words[first_letter] = [word]
        else:
            grouped_words[first_letter].append(word)
    return grouped_words
```
- **Análise:** A função percorre todos os `n` elementos na lista de palavras e agrupa cada palavra com base na primeira letra.
- **Complexidade no pior caso:** \(O(n)\), onde `n` é o número de palavras na lista.

---

### **Funções Trabalhando com Conjuntos**

#### **Alternativa a: Operações em Conjuntos**
```python
def set_operations(set1, set2):
    union_set = set1 | set2
    intersection_set = set1 & set2
    difference_set1 = set1 - set2
    difference_set2 = set2 - set1
    return {"Union": union_set, "Intersection": intersection_set, "Difference (Set 1 - Set 2)": difference_set1, "Difference

 (Set 2 - Set 1)": difference_set2}
```
- **Análise:** As operações de união, interseção e diferença entre dois conjuntos são \(O(\min(|A|, |B|))\), onde `A` e `B` são os conjuntos.
- **Complexidade no pior caso:** \(O(n + m)\), onde `n` e `m` são os tamanhos dos dois conjuntos.

#### **Alternativa b: Verificar Subconjunto**
```python
def check_subset(set1, set2):
    return set1.issubset(set2)
```
- **Análise:** A operação de verificação de subconjunto é \(O(\min(|A|, |B|))\), onde `A` e `B` são os conjuntos.
- **Complexidade no pior caso:** \(O(\min(n, m))\), onde `n` e `m` são os tamanhos dos conjuntos.

#### **Alternativa c: Calcular Diferença Simétrica**
```python
def symmetric_difference(set1, set2):
    return set1 ^ set2
```
- **Análise:** A operação de diferença simétrica entre dois conjuntos é \(O(\min(|A|, |B|))\).
- **Complexidade no pior caso:** \(O(n + m)\), onde `n` e `m` são os tamanhos dos conjuntos.

#### **Alternativa d: Contagem de Elementos**
```python
def count_elements_in_set(set1):
    return len(set1)
```
- **Análise:** A operação de contar o número de elementos em um conjunto é \(O(1)\), pois a implementação de conjuntos em Python mantém esse valor armazenado.
- **Complexidade no pior caso:** \(O(1)\).

### Questão 5

#### **Analisando a Complexidade de Tempo da Função**

A função dada é:

```python
def mystery_function(n):
    total = 0
    for i in range(n):
        for j in range(i):
            total += j
    return total
```

#### **Loop Externo (`i`):**
- O loop externo percorre de `i = 0` até `i = n - 1`, ou seja, ele roda **n** vezes. Isso nos dá uma complexidade de **O(n)** para o loop externo.

#### **Loop Interno (`j`):**
- O loop interno percorre de `j = 0` até `j = i - 1`. Assim, para cada iteração do loop externo (onde `i` é a variável de iteração), o loop interno executa **i** vezes.

#### **Total de Iterações:**
O número total de operações (o total de iterações do loop interno) pode ser expresso como a soma dos números inteiros de 0 até `n-1`:

$$
0 + 1 + 2 + \dots + (n-1)
$$

Essa soma é uma **série aritmética**. O valor dessa soma é dado pela fórmula:

$$
\text{Soma} = \frac{(n-1) \cdot n}{2}
$$

Simplificando a expressão para o comportamento assintótico:

$$
O(n^2)
$$

#### **Conclusão:**

A complexidade de tempo no pior caso é dominada pelos loops aninhados. Como o número total de operações é proporcional à soma dos primeiros `n-1` números inteiros, a complexidade de tempo é **O(n^2)**. Portanto, a **complexidade de tempo no pior caso** da função `mystery_function(n)` é:

$$
O(n^2)
$$